0000-pre-shared-key-v2.md

```diff
@@ -0,0 +1,324 @@
+2. Sender generates the payment condition as described [below](#payment-chunk-condition).
+3. Sender sends a payment with the condition from step 2., the encrypted request from step 1. as the data, and an arbitrary source amount.
+4. Receiver is notified of the incoming transfer and decrypts the data to determine it is for a payment chunk.
+5. Receiver creates a [Payment Chunk Response](#payment-chunk-response) with the same Payment ID, Sequence, and Payment Received amount. Receiver sets the Chunk Received Amount to the amount that arrived on the incoming transfer. Receiver [encrypts the response](#encryption-envelope).
+6. Receiver generates fulfillment as described [below](#payment-chunk-fulfillment) and fulfills the incoming transfer.
+7. Sender is notified of the outgoing transfer being fulfilled and decrypts the attached data.
+8. Sender determines the exchange rate by dividing the Chunk Amount from the receiver by the chunk's source amount.
+9. Sender repeats step 1-8 until the fixed source or destination amount has been sent or delivered, respectively. Sender MAY adjust the chunk amount upwards or downwards in response to successful and unsuccessful payments.
+
+## Specification
+
+### Encryption
+
+Request and response data are encrypted using [AES-GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) with a 16-byte Initialization Vector (IV).
+
+#### Encryption Envelope
+
+```
+(Numbers represent bytes)
+
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|           Random IV           |        Auth Tag       | Ciph..|
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          Ciphertext                           |
+\                                                               \
+\                                                               \
+|                          Ciphertext                           |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+```
+
+| Field | Type | Description |
+|---|---|---|
+| Random IV  | 16-Byte UInt | Nonce used as input to the AES-GCM algorithm. Also ensures conditions are random |
+| Auth Tag   | 12-Byte UInt | Authentication tag produced by AES-GCM encryption that ensures data integrity |
+| Ciphertext | 0-32739 Bytes | Encrypted data (see below for contents) |
+
+#### Encryption Pseudocode
+
+Note that the encryption key is derived from the shared secret using an HMAC.
+
+```
+iv = random_bytes(16)
+encryption_key = hmac_sha256(shared_secret, "ilp_psk_encryption")
+{ ciphertext, auth_tag } = aes_256_gcm(encryption_key, iv, data)
+```
+
+### Quotes
+
+#### Quote Request
```

Comment:
@justmoon brought up the point: do we need purely informational quotes or should these be removed?

If on a normal payment the receiver will tell you how much arrived, the sender could just use an actual payment to determine the rate for the path. It seems like the main problem would be that the rounding error could get quite large if you send a super tiny payment first. That would also mean you need to pay a small amount for quotes, but that might not be the worst thing.

Thoughts?

Reply:
If a quote is a "Payment Chunk Request" with the "Chunk Minimum" set to the max uint64, then it won't actually cost anything.

Reply:
That's a great point. I'm going to update it to remove the quote request and response types. It's very cool that that means we can use the same request/response/error packets to handle quotes, one-off payments, and chunked payments.